File: zinc-flow-csharp/ZincFlow/Fabric/SchemaRegistryHandler.cs
using System.Text;
using System.Text.Json;
using ZincFlow.Core;
using ZincFlow.StdLib;

namespace ZincFlow.Fabric;

/// <summary>
/// Confluent-shape REST endpoints for the embedded schema registry.
/// Mounted under /api/schema-registry/* on the management port.
///
/// JSON parsed/built with hand-rolled Utf8JsonReader/Writer to stay AOT-clean —
/// no DTO classes registered in ZincJsonContext.
/// 4xx response bodies match Confluent's error_code/message shape so any
/// existing tooling that handles those errors works against us.
/// </summary>
public sealed class SchemaRegistryHandler
{
    private readonly EmbeddedSchemaRegistry _registry;

    public SchemaRegistryHandler(EmbeddedSchemaRegistry registry) => _registry = registry;

    public void MapRoutes(WebApplication app)
    {
        app.MapGet("/api/schema-registry/schemas/ids/{id:int}", GetById);
        app.MapGet("/api/schema-registry/subjects", ListSubjects);
        app.MapGet("/api/schema-registry/subjects/{subject}/versions", ListVersions);
        app.MapGet("/api/schema-registry/subjects/{subject}/versions/{version}", GetSubjectVersion);
        app.MapPost("/api/schema-registry/subjects/{subject}/versions", (Delegate)RegisterSchema);
        app.MapDelete("/api/schema-registry/subjects/{subject}", DeleteSubject);
        app.MapDelete("/api/schema-registry/subjects/{subject}/versions/{version:int}", DeleteVersion);
    }

    // --- GET /schemas/ids/{id} → {"schema":"..."} ---

    private IResult GetById(int id)
    {
        var entry = _registry.GetEntryByIdInternal(id);
        if (entry is null) return JsonError(404, 40403, $"schema id {id} not found");
        return JsonContent(BuildJson(("schema", AvroSchemaJson.Emit(entry.Value.Schema))));
    }

    // --- GET /subjects → ["a","b","c"] ---

    private IResult ListSubjects() => JsonContent(BuildStringArray(_registry.ListSubjects()));

    // --- GET /subjects/{subject}/versions → [1,2,3] ---

    private IResult ListVersions(string subject)
    {
        var versions = _registry.ListVersions(subject);
        if (versions.Count == 0) return JsonError(404, 40401, $"subject '{subject}' not found");
        return JsonContent(BuildIntArray(versions));
    }

    // --- GET /subjects/{subject}/versions/{version|latest} → full payload ---

    private IResult GetSubjectVersion(string subject, string version)
    {
        EmbeddedSchemaRegistry.VersionEntry entry;
        if (version == "latest")
        {
            if (!_registry.TryGetLatest(subject, out entry))
                return JsonError(404, 40401, $"subject '{subject}' not found");
        }
        else if (int.TryParse(version, out var v))
        {
            if (!_registry.TryGetEntry(subject, v, out entry))
                return JsonError(404, 40402, $"subject '{subject}' version {v} not found");
        }
        else
        {
            return JsonError(400, 40000, "version must be an integer or 'latest'");
        }
        return JsonContent(BuildJson(
            ("subject", subject),
            ("version", entry.Version),
            ("id", entry.Id),
            ("schema", AvroSchemaJson.Emit(entry.Schema))));
    }

    // --- POST /subjects/{subject}/versions  body {"schema":"..."}  → {"id":N} ---

    private async Task<IResult> RegisterSchema(string subject, HttpContext ctx)
    {
        string body;
        using (var sr = new StreamReader(ctx.Request.Body, Encoding.UTF8))
            body = await sr.ReadToEndAsync();

        var schemaJson = ExtractSchemaField(body);
        if (schemaJson is null) return JsonError(422, 42201, "request body missing 'schema' string field");

        Schema schema;
        try { schema = AvroSchemaJson.Parse(schemaJson); }
        catch (Exception ex) { return JsonError(422, 42202, $"invalid Avro schema: {ex.Message}"); }

        int id;
        try { id = await _registry.RegisterAsync(subject, schema); }
        catch (Exception ex) { return JsonError(500, 50001, $"register failed: {ex.Message}"); }

        return JsonContent(BuildJson(("id", id)));
    }

    // --- DELETE /subjects/{subject} → list of deleted versions ---

    private IResult DeleteSubject(string subject)
    {
        var versions = _registry.ListVersions(subject);
        if (versions.Count == 0) return JsonError(404, 40401, $"subject '{subject}' not found");
        _registry.DeleteSubject(subject);
        return JsonContent(BuildIntArray(versions));
    }

    // --- DELETE /subjects/{subject}/versions/{version} → version deleted ---

    private IResult DeleteVersion(string subject, int version)
    {
        if (!_registry.DeleteVersion(subject, version))
            return JsonError(404, 40402, $"subject '{subject}' version {version} not found");
        return JsonContent(version.ToString());
    }

    // --- Hand-rolled JSON helpers ---

    private static IResult JsonContent(string json)
        => Results.Content(json, "application/vnd.schemaregistry.v1+json");

    private static IResult JsonError(int statusCode, int errorCode, string message)
    {
        var body = BuildJson(("error_code", errorCode), ("message", message));
        return Results.Content(body, "application/vnd.schemaregistry.v1+json", statusCode: statusCode);
    }

    private static string BuildJson(params (string Key, object Value)[] pairs)
    {
        using var ms = new MemoryStream();
        using (var w = new Utf8JsonWriter(ms))
        {
            w.WriteStartObject();
            foreach (var (k, v) in pairs)
            {
                switch (v)
                {
                    case int i: w.WriteNumber(k, i); break;
                    case long l: w.WriteNumber(k, l); break;
                    case string s: w.WriteString(k, s); break;
                    case bool b: w.WriteBoolean(k, b); break;
                    default: w.WriteString(k, v.ToString() ?? ""); break;
                }
            }
            w.WriteEndObject();
        }
        return Encoding.UTF8.GetString(ms.ToArray());
    }

    private static string BuildStringArray(List<string> items)
    {
        using var ms = new MemoryStream();
        using (var w = new Utf8JsonWriter(ms))
        {
            w.WriteStartArray();
            foreach (var s in items) w.WriteStringValue(s);
            w.WriteEndArray();
        }
        return Encoding.UTF8.GetString(ms.ToArray());
    }

    private static string BuildIntArray(List<int> items)
    {
        using var ms = new MemoryStream();
        using (var w = new Utf8JsonWriter(ms))
        {
            w.WriteStartArray();
            foreach (var i in items) w.WriteNumberValue(i);
            w.WriteEndArray();
        }
        return Encoding.UTF8.GetString(ms.ToArray());
    }

    /// <summary>Extracts the "schema" string field from a Confluent register body.</summary>
    private static string? ExtractSchemaField(string body)
    {
        var bytes = Encoding.UTF8.GetBytes(body);
        var reader = new Utf8JsonReader(bytes);
        while (reader.Read())
        {
            if (reader.TokenType != JsonTokenType.PropertyName) continue;
            if (reader.GetString() != "schema") continue;
            reader.Read();
            return reader.TokenType == JsonTokenType.String ? reader.GetString() : null;
        }
        return null;
    }
}

File: zinc-flow-java/src/main/java/zincflow/fabric/SchemaRegistryHandler.java
package zincflow.fabric;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.javalin.Javalin;
import io.javalin.http.Context;
import zincflow.providers.SchemaRegistryProvider;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/// Confluent-shape REST surface over {@link SchemaRegistryProvider}.
/// Mounted under {@code /api/schema-registry/*} alongside the rest of
/// the management API so serializer clients that speak to Confluent
/// Schema Registry work verbatim against an embedded zinc-flow-java.
///
/// <h2>Mapped routes</h2>
/// <pre>
///   GET    /schemas/ids/{id}
///   GET    /subjects
///   GET    /subjects/{subject}/versions
///   GET    /subjects/{subject}/versions/{version|latest}
///   POST   /subjects/{subject}/versions          body: {"schema":"..."}
///   DELETE /subjects/{subject}
///   DELETE /subjects/{subject}/versions/{version}
/// </pre>
///
/// Responses carry
/// {@code Content-Type: application/vnd.schemaregistry.v1+json}, and
/// 4xx/5xx bodies match Confluent's {@code {"error_code", "message"}}
/// shape so any existing tooling that handles those errors works
/// against us unchanged. Mirrors zinc-flow-csharp's
/// {@code SchemaRegistryHandler}.
public final class SchemaRegistryHandler {

    public static final String CONTENT_TYPE = "application/vnd.schemaregistry.v1+json";

    private final SchemaRegistryProvider registry;
    private final ObjectMapper json = new ObjectMapper();

    public SchemaRegistryHandler(SchemaRegistryProvider registry) {
        if (registry == null) throw new IllegalArgumentException("registry must not be null");
        this.registry = registry;
    }

    public void mapRoutes(Javalin app) {
        app.get("/api/schema-registry/schemas/ids/{id}",                    this::getById);
        app.get("/api/schema-registry/subjects",                            this::listSubjects);
        app.get("/api/schema-registry/subjects/{subject}/versions",         this::listVersions);
        app.get("/api/schema-registry/subjects/{subject}/versions/{version}", this::getVersion);
        app.post("/api/schema-registry/subjects/{subject}/versions",        this::register);
        app.delete("/api/schema-registry/subjects/{subject}",               this::deleteSubject);
        app.delete("/api/schema-registry/subjects/{subject}/versions/{version}", this::deleteVersion);
    }

    // --- GET /schemas/ids/{id} → {"schema":"..."} ---

    private void getById(Context ctx) throws Exception {
        int id;
        try { id = Integer.parseInt(ctx.pathParam("id")); }
        catch (NumberFormatException e) {
            error(ctx, 400, 40000, "id must be an integer");
            return;
        }
        var entry = registry.getById(id);
        if (entry.isEmpty()) {
            error(ctx, 404, 40403, "schema id " + id + " not found");
            return;
        }
        write(ctx, 200, Map.of("schema", entry.get().definition()));
    }

    // --- GET /subjects → ["a","b","c"] ---

    private void listSubjects(Context ctx) throws Exception {
        writeRaw(ctx, 200, json.writeValueAsBytes(registry.listSubjects()));
    }

    // --- GET /subjects/{subject}/versions → [1,2,3] ---

    private void listVersions(Context ctx) throws Exception {
        String subject = ctx.pathParam("subject");
        List<Integer> versions = registry.listVersions(subject);
        if (versions.isEmpty()) {
            error(ctx, 404, 40401, "subject '" + subject + "' not found");
            return;
        }
        writeRaw(ctx, 200, json.writeValueAsBytes(versions));
    }

    // --- GET /subjects/{subject}/versions/{version|latest} → full payload ---

    private void getVersion(Context ctx) throws Exception {
        String subject = ctx.pathParam("subject");
        String version = ctx.pathParam("version");
        var entry = "latest".equals(version)
                ? registry.latest(subject)
                : parseVersion(ctx, version).flatMap(v -> registry.getEntry(subject, v));
        if (entry.isEmpty()) {
            // Distinguish unknown subject vs unknown version — Confluent does.
            if (!registry.listSubjects().contains(subject)) {
                error(ctx, 404, 40401, "subject '" + subject + "' not found");
            } else {
                error(ctx, 404, 40402, "subject '" + subject + "' version " + version + " not found");
            }
            return;
        }
        var s = entry.get();
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("subject", s.subject());
        body.put("version", s.version());
        body.put("id", s.id());
        body.put("schema", s.definition());
        write(ctx, 200, body);
    }

    private java.util.Optional<Integer> parseVersion(Context ctx, String raw) {
        try {
            return java.util.Optional.of(Integer.parseInt(raw));
        } catch (NumberFormatException e) {
            return java.util.Optional.empty();
        }
    }

    // --- POST /subjects/{subject}/versions  body {"schema":"..."}  → {"id":N} ---

    private void register(Context ctx) throws Exception {
        String subject = ctx.pathParam("subject");
        String body = ctx.body();
        String schemaDef;
        try {
            JsonNode root = json.readTree(body);
            JsonNode schemaNode = root.get("schema");
            if (schemaNode == null || !schemaNode.isTextual()) {
                error(ctx, 422, 42201, "request body missing 'schema' string field");
                return;
            }
            schemaDef = schemaNode.asText();
        } catch (Exception ex) {
            error(ctx, 422, 42201, "request body is not valid JSON: " + ex.getMessage());
            return;
        }

        try {
            var registered = registry.register(subject, schemaDef);
            write(ctx, 200, Map.of("id", registered.id()));
        } catch (IllegalArgumentException ex) {
            error(ctx, 422, 42202, "register failed: " + ex.getMessage());
        } catch (RuntimeException ex) {
            error(ctx, 500, 50001, "register failed: " + ex.getMessage());
        }
    }

    // --- DELETE /subjects/{subject} → [1,2,3] of removed versions ---

    private void deleteSubject(Context ctx) throws Exception {
        String subject = ctx.pathParam("subject");
        List<Integer> removed = registry.deleteSubject(subject);
        if (removed.isEmpty()) {
            error(ctx, 404, 40401, "subject '" + subject + "' not found");
            return;
        }
        writeRaw(ctx, 200, json.writeValueAsBytes(removed));
    }

    // --- DELETE /subjects/{subject}/versions/{version} → version number ---

    private void deleteVersion(Context ctx) throws Exception {
        String subject = ctx.pathParam("subject");
        int version;
        try { version = Integer.parseInt(ctx.pathParam("version")); }
        catch (NumberFormatException e) {
            error(ctx, 400, 40000, "version must be an integer");
            return;
        }
        boolean deleted = registry.deleteVersion(subject, version);
        if (!deleted) {
            error(ctx, 404, 40402, "subject '" + subject + "' version " + version + " not found");
            return;
        }
        writeRaw(ctx, 200, json.writeValueAsBytes(version));
    }

    // --- helpers ---

    private void write(Context ctx, int status, Object body) throws Exception {
        ctx.status(status)
           .contentType(CONTENT_TYPE)
           .result(json.writeValueAsBytes(body));
    }

    private void writeRaw(Context ctx, int status, byte[] body) {
        ctx.status(status).contentType(CONTENT_TYPE).result(body);
    }

    private void error(Context ctx, int status, int errorCode, String message) throws Exception {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("error_code", errorCode);
        body.put("message", message);
        write(ctx, status, body);
    }
}
